### PR TITLE
[telemetry] allow to filter logs using zapfilter rules or a regular expression

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
Add 2 new config options for logs:
* `filter_rules`: taken after the zapfilter library, allows to apply filtering to logs.
* `filter_regex`: a regular regular expression. Any matched log (message only) is excluded.

**Link to tracking Issue:**
#7107

**Testing:**
Unit tests.

**Documentation:**
Added to go docs.

### CODE REVIEW

**Comment by atoulme** [comment]
This is built on top of #7094, which should be looked at first.

**Comment by dmitryax** [comment]
Please update the issue link, it points to the PR itself.

**Comment by atoulme** [comment]
I assumed that was ok based on [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18205#issuecomment-1410974587) - will fix.

**Comment by dmitryax** [comment]
> I assumed that was ok based on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18205#issuecomment-1410974587 - will fix.

That link is in the changelog, not PR description. This change introduces a bunch of public API that we are trying to reduce and stabilize for 1.0, so we need to understand the problem that is intended to be solved by this.

**Comment by atoulme** [comment]
No worries. I have created the issue and updated links.

**Comment by atoulme** [comment]
Closing as per SIG discussion, we will use #7223's approach instead.

**Comment by atoulme** [comment]
Reopening since https://github.com/open-telemetry/opentelemetry-collector/issues/7223 is abandoned.

File: service/telemetry/config.go [hunk 1]
```diff
@@ -94,6 +94,23 @@ type LogsConfig struct {
...
 	InitialFields map[string]any `mapstructure:"initial_fields"`
+
+	// Filtering rules to exclude log statements. See https://github.com/moul/zapfilter
+	// for documentation.
+	//
+	// Example:
+	//
+	//      filter_rules: "*:myns info,warn:myns.* error:*"
+	// By default, there is no filter.
+	FilterRules string `mapstructure:"filter_rules"`
```
Comment (dmitryax) [inline service/telemetry/config.go:105]: This field name seems a bit unclear, and we don't have a pattern like this in other collector configurations. The closest one is stanza include/exclude https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver#configuration. Maybe we adopt the same `exclude` name here?

File: service/telemetry/config.go [hunk 2]
```diff
@@ -94,6 +94,23 @@ type LogsConfig struct {
...
+
+	// Filtering regular expression to exclude log statements.
+	//
+	// Example:
+	//
+	//      filter_regex: "partition.*"
+	// By default, there is no regular expression filter.
+	FilterRegex string `mapstructure:"filter_regex"`
```
Comment (dmitryax) [inline service/telemetry/config.go:113]: Do we need regexp? Can we keep the blob only for now? If we will need regexp, we can apply the same approach as in the signalfx exporter and read values like `/regexp/` from the same `exclude` field.

Comment (atoulme) [reply service/telemetry/config.go:113]: We don't have a blob approach right now, did I miss something?

File: service/telemetry/config.go (hunk 1 above)
Comment (atoulme) [reply service/telemetry/config.go:105]: This is exposing a zapfilter ruleset. From their [README](https://github.com/moul/zapfilter):
```
*                            everything
        *:*                          everything
        info:*                       level info;  any namespace
        info+:*                      levels info, warn, error, dpanic, panic, and fatal; any namespace
        info,warn:*                  levels info, warn; any namespace
        ns1                          any level; namespace 'ns1'
        *:ns1                        any level; namespace 'ns1'
        ns1*                         any level; namespaces matching 'ns1*'
        *:ns1*                       any level; namespaces matching 'ns1*'
        *:ns1,ns2                    any level; namespaces 'ns1' and 'ns2'
        *:ns*,-ns3*                  any level; namespaces matching 'ns*' but not matching 'ns3*'
        info:ns1                     level info; namespace 'ns1'
        info,warn:ns1,ns2            levels info and warn; namespaces 'ns1' and 'ns2'
        info:ns1 warn:n2             level info + namespace 'ns1' OR level warn and namespace 'ns2'
        info,warn:myns* error+:*     levels info or warn and namespaces matching 'myns*' OR levels error, dpanic, panic or fatal for any namespace
```

This is not a glob approach - this is specifically allowing the namespaces and levels set in the rules.

**Comment by dwalczakfx** [comment]
Can we reopen this ?

**Pull request closed without merging**